File: leetcode_ly/java/Array/3SumClosest.java
public class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if(nums.length < 3) return target;
        int res = Integer.MAX_VALUE;
        int sum = Math.abs(nums[0] + nums[1] + nums[2] - target);
        quickSort(nums,0,nums.length - 1);
        int lastNum = Integer.MAX_VALUE;
        for(int i = 0; i < nums.length - 1; i++){
            if(nums[i] == lastNum) continue;
            int j = i + 1;
            int k = nums.length - 1;
            while(j < k){
                if(nums[i] + nums[j] + nums[k] == target) return target;
                else if(nums[i] + nums[j] + nums[k] < target){
                    sum = (res > target - (nums[i] + nums[j] + nums[k])) ? nums[i] + nums[j] + nums[k] : sum;
                    res = Math.min(res,target - (nums[i] + nums[j] + nums[k]));
                    j++;
                }
                else{
                    sum = (res > nums[i] + nums[j] + nums[k] - target) ? nums[i] + nums[j] + nums[k] : sum;
                    res = Math.min(res,nums[i] + nums[j] + nums[k] - target);
                    k--;
                }
            }
            lastNum = nums[i];
        }
        return sum;
    }
    
    public void quickSort(int[] nums,int left,int right){
        if(left >= right) return;
        int low = left;
        int high = right;
        int target = nums[low];
        while(low < high){
            while(low < high && nums[high] >= target){
                high--;
            }
            nums[low] = nums[high];
            while(low < high && nums[low] <= target){
                low++;
            }
            nums[high] = nums[low];
        }
        nums[low] = target;
        quickSort(nums,left,low - 1);
        quickSort(nums,low + 1,right);
    } 
}

File: leetcode_ly/c++/Array/3SumClosest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        if(nums.size() < 3){
            return target;
        }
        else{
            vector<int> b = sort(nums);
            int len = b.size();
            int diff = b[0] + b[1] +b[2] - target;
            vector<int> result;
            result.push_back(b[0]);
            result.push_back(b[1]);
            result.push_back(b[2]);
            int currentValue = b[0];
            for(int i = 0; i < len - 1; i++){
                if(currentValue == b[i] && i){
                    continue;
                }
                else{
                    int begin = i + 1;
                    int end = len - 1;
                    while(begin < end){
                        int sum = b[i] + b[begin] + b[end] - target;
                        if(abs(sum) <= abs(diff) && sum < 0){
                            diff = abs(sum);
                            result[0] = b[i];
                            result[1] = b[begin];
                            result[2] = b[end];
                            begin++;
                        }
                        else if(abs(sum) < abs(diff) && sum > 0){
                            diff = abs(sum);
                            result[0] = b[i];
                            result[1] = b[begin];
                            result[2] = b[end];
                            end--;
                        }
                        else if(sum == 0){
                            return b[i] + b[begin] + b[end];
                        }
                        else if(abs(sum) > abs(diff) && sum < 0){
                            begin++;
                        }
                        else{
                        	end--;
						}
                    }
                   
                }
                currentValue = b[i];
            }
            return result[0] + result[1] + result[2];  
        }
    }
   
    vector<int> sort(vector<int>& nums){
        int len = nums.size();
        for(int i = 0; i < len - 1; i++){
            for(int j = i + 1; j < len; j++){
                if(nums[i] >= nums[j]){
                    int temp = nums[i];
                    nums[i] = nums[j];
                    nums[j] = temp;
                }
            }
        }
        return nums;
    }
};